txpool/txpool.go

```diff
@@ -346,63 +338,46 @@ func (p *TxPool) Pop(tx *types.Transaction) {
 	}
 }
 
-// Drop removes the given (unrecoverable) transaction from
-// its associated promoted queue (account) and rolls
-// back the account's nextNonce.
-// Will update executables with the next primary
-// from that account (if any).
+// Drop clears the entire account associated with the given transaction
+// and reverts its next (expected) nonce.
 func (p *TxPool) Drop(tx *types.Transaction) {
+	// fetch associated account
 	account := p.accounts.get(tx.From)
 
 	account.promoted.lock(true)
-	defer account.promoted.unlock()
+	account.enqueued.lock(true)
 
-	// pop the top most promoted tx
-	account.promoted.pop()
+	defer func() {
+		account.enqueued.unlock()
+		account.promoted.unlock()
+	}()
 
-	// update state
-	p.index.remove(tx)
-	p.gauge.decrease(slotsRequired(tx))
+	// rollback nonce
+	nextNonce := tx.Nonce
+	account.setNonce(nextNonce)
 
-	// update metrics
-	p.metrics.PendingTxs.Add(-1)
+	// drop promoted
+	dropped := account.promoted.clear()
+	p.index.remove(dropped...)
+	p.gauge.decrease(slotsRequired(dropped...))
```

Comment:
Can we extract these into a helper method? (could be inside `Drop`, since we use it a few lines further)

Reply:
Sure, just give me a clever name for that helper method.